dgram: skip dns lookup for IPs

evanlucas opened the pull request:
Previously, we would call dns.lookup even for IP addresses
in Socket#connect. This now skips that to improve performance.

Fixes: https://github.com/nodejs/node/issues/39468

evanlucas committed:
dgram: skip dns lookup for IPs

Previously, we would call dns.lookup even for IP addresses
in Socket#connect. This now skips that to improve performance.

Fixes: https://github.com/nodejs/node/issues/39468

schmod commented:
I'd avoided PRing this myself, because I couldn't figure out the best way to share `isIP` between `net` and `dgram` – I don't think we want `dgram` to depend on `net`, because it seems like it could introduce subtle changes to the order that modules load, which seems like it could break many APM implementations.

evanlucas commented:
@schmod, good point, I got a rebase conflict somehow :/ so have to fix that regardless. With it loading internal/net for `isIP`, I'm not sure that would change anything as far as module load time?

schmod commented:
> With it loading internal/net for isIP, I'm not sure that would change anything as far as module load time?

True, I'm not really worried about that.  It's more about whether or not anybody's hot-patching any modules used downstream of `net`.  

That all being said, that sounds like an extreme edge-case, and might not be worth worrying about?

jasnell commented:
While it may seem odd, this may be a breaking (semver-major) change. User code is permitted to provide a custom dns lookup function when the dgram/udp socket is created. Prior to this change, that function would be called for IP address values.

jasnell commented:
This should include (a) a test case that verifies that the lookup function is not called and (b) a doc update indicating that a custom lookup function will not be invoked when the address is an IP address.

schmod commented:
> While it may seem odd, this may be a breaking (semver-major) change. User code is permitted to provide a custom dns lookup function when the dgram/udp socket is created. Prior to this change, that function would be called for IP address values.

We could preserve the current behavior if a custom `lookup` is provided, and postpone the breaking change to the next major release...

evanlucas committed:
fixup!: make work and start adding tests

They are currently failing due to us not emitting the
appropriate addressType in the lookup event.

JackFreelanderHBO commented:
@evanlucas - Would it be possible to revive this PR and complete the fix?  There seems to be no work-around for dgram's socket implementation invoking DNS resolution on every single `.send()` call without this fix.  It would be great to have this.
